File: kattis/longest-palindromic-substring/longest-palindromic-substring.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    vector<vector<int>> memo;
    string s;
    
    int dp(int start, int end) { // end inclusive
        if (start == end) return true;
        if (abs(start-end) == 1) return s[start] == s[end];
        int &ans = memo[start][end];
        if (ans != -1) return ans;
        return ans = (s[start] == s[end]) && dp(start+1, end-1);
    }
public:
    string longestPalindrome(string ss) {
        int MAX_N = 1010;
        memo = vector<vector<int>>(MAX_N, vector<int>(MAX_N, -1));
        s = ss;
        
        //dp(0, s.size()-1);
        
        //cout << memo[1][4] << endl;
        
        int start = 0, end = 0;
        for (int i = 0; i < s.size(); i++) {
            for (int j = i; j < s.size(); j++) {
                if (dp(i, j) && ((j-i+1) >= (end-start+1))) {
                    start = i;
                    end = j;
                }
            }
        }
        
        return s.substr(start, end-start+1);
    }
};

File: kattis/longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        ans = -1
        ansS = ""
        
        def expand(i, j):
            if i < 0 or j >= len(s) or s[i] != s[j]: 
                return -1, ""
            length = 1 if i == j else 2
            i -= 1
            j += 1
            while i >= 0 and j < len(s):
                if s[i] == s[j]:
                    length += 2
                else:
                    break
                i -= 1
                j += 1
            return length, s[i+1:j]
        
        for i in range(len(s)):
            ans1, s1 = expand(i, i)
            if ans1 > ans:
                ans = ans1
                ansS = s1
            ans2, s2 = expand(i, i+1)
            if ans2 > ans:
                ans = ans2
                ansS = s2
        
        return ansS